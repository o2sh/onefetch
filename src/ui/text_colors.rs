use crate::ui::num_to_color;
use owo_colors::{AnsiColors, DynColors};

#[derive(Clone)]
pub struct TextColors {
    pub title: DynColors,
    pub tilde: DynColors,
    pub underline: DynColors,
    pub subtitle: DynColors,
    pub colon: DynColors,
    pub info: DynColors,
}

impl TextColors {
    pub fn new(colors: &[u8], logo_primary_color: DynColors) -> Self {
        let mut text_colors = Self {
            title: logo_primary_color,
            tilde: DynColors::Ansi(AnsiColors::Default),
            underline: DynColors::Ansi(AnsiColors::Default),
            subtitle: logo_primary_color,
            colon: DynColors::Ansi(AnsiColors::Default),
            info: DynColors::Ansi(AnsiColors::Default),
        };

        if !colors.is_empty() {
            let custom_color = colors.iter().map(num_to_color).collect::<Vec<DynColors>>();

            text_colors.title = *custom_color.first().unwrap_or(&logo_primary_color);
            text_colors.tilde = *custom_color
                .get(1)
                .unwrap_or(&DynColors::Ansi(AnsiColors::Default));
            text_colors.underline = *custom_color
                .get(2)
                .unwrap_or(&DynColors::Ansi(AnsiColors::Default));
            text_colors.subtitle = *custom_color.get(3).unwrap_or(&logo_primary_color);
            text_colors.colon = *custom_color
                .get(4)
                .unwrap_or(&DynColors::Ansi(AnsiColors::Default));
            text_colors.info = *custom_color
                .get(5)
                .unwrap_or(&DynColors::Ansi(AnsiColors::Default));
        }
        text_colors
    }
}

#[cfg(test)]
mod test {
    use super::*;

    #[test]
    fn no_custom_colors() {
        let primary_color = DynColors::Ansi(AnsiColors::Blue);
        let text_colors = TextColors::new(&[], primary_color);
        assert_eq!(text_colors.title, primary_color);
        assert_eq!(text_colors.tilde, DynColors::Ansi(AnsiColors::Default));
        assert_eq!(text_colors.underline, DynColors::Ansi(AnsiColors::Default));
        assert_eq!(text_colors.subtitle, primary_color);
        assert_eq!(text_colors.colon, DynColors::Ansi(AnsiColors::Default));
        assert_eq!(text_colors.info, DynColors::Ansi(AnsiColors::Default));
    }

    #[test]
    fn with_custom_colors() {
        let custom_colors = vec![0, 1, 2, 3, 4, 5];
        let text_colors = TextColors::new(&custom_colors, DynColors::Ansi(AnsiColors::Blue));
        assert_eq!(text_colors.title, num_to_color(&custom_colors[0]));
        assert_eq!(text_colors.tilde, num_to_color(&custom_colors[1]));
        assert_eq!(text_colors.underline, num_to_color(&custom_colors[2]));
        assert_eq!(text_colors.subtitle, num_to_color(&custom_colors[3]));
        assert_eq!(text_colors.colon, num_to_color(&custom_colors[4]));
        assert_eq!(text_colors.info, num_to_color(&custom_colors[5]));
    }

    #[test]
    fn with_some_custom_colors() {
        let custom_colors = vec![0, 1, 2];
        let primary_color = DynColors::Ansi(AnsiColors::Blue);
        let text_colors = TextColors::new(&custom_colors, primary_color);
        assert_eq!(text_colors.title, num_to_color(&custom_colors[0]));
        assert_eq!(text_colors.tilde, num_to_color(&custom_colors[1]));
        assert_eq!(text_colors.underline, num_to_color(&custom_colors[2]));
        assert_eq!(text_colors.subtitle, primary_color);
        assert_eq!(text_colors.colon, DynColors::Ansi(AnsiColors::Default));
        assert_eq!(text_colors.info, DynColors::Ansi(AnsiColors::Default));
    }
}
