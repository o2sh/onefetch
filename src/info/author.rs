use super::git::metrics::GitMetrics;
use crate::{
    cli::NumberSeparator,
    info::{format_number, utils::info_field::InfoField},
};
use serde::Serialize;
use std::fmt::Write;

#[derive(Serialize, Clone, Debug, PartialEq)]
#[serde(rename_all = "camelCase")]
pub struct Author {
    pub name: String,
    email: Option<String>,
    nbr_of_commits: usize,
    contribution: usize,
    #[serde(skip_serializing)]
    number_separator: NumberSeparator,
}

impl Author {
    pub fn new(
        name: String,
        email: Option<String>,
        nbr_of_commits: usize,
        total_nbr_of_commits: usize,
        number_separator: NumberSeparator,
    ) -> Self {
        let contribution =
            (nbr_of_commits as f32 * 100. / total_nbr_of_commits as f32).round() as usize;
        Self {
            name,
            email,
            nbr_of_commits,
            contribution,
            number_separator,
        }
    }
}

impl std::fmt::Display for Author {
    fn fmt(&self, f: &mut std::fmt::Formatter) -> std::fmt::Result {
        if let Some(email) = &self.email {
            write!(
                f,
                "{}% {} <{}> {}",
                self.contribution,
                self.name,
                email,
                format_number(&self.nbr_of_commits, self.number_separator)
            )
        } else {
            write!(
                f,
                "{}% {} {}",
                self.contribution,
                self.name,
                format_number(&self.nbr_of_commits, self.number_separator)
            )
        }
    }
}

#[derive(Serialize)]
pub struct AuthorsInfo {
    pub authors: Vec<Author>,
}

impl AuthorsInfo {
    pub fn new(git_metrics: &GitMetrics) -> Self {
        let authors = git_metrics.authors_to_display.clone();
        Self { authors }
    }

    fn top_contribution(&self) -> usize {
        if let Some(top_contributor) = self.authors.get(0) {
            return top_contributor.contribution;
        }
        0
    }
}

fn digit_difference(num1: usize, num2: usize) -> usize {
    let count_digits = |num: usize| (num.checked_ilog10().unwrap_or(0) + 1) as usize;

    count_digits(num1) - count_digits(num2)
}

impl std::fmt::Display for AuthorsInfo {
    fn fmt(&self, f: &mut std::fmt::Formatter) -> std::fmt::Result {
        let mut authors_info = String::new();

        let pad = self.title().len() + 2;
        for (i, author) in self.authors.iter().enumerate() {
            if i == 0 {
                write!(authors_info, "{author}")?;
            } else {
                write!(
                    authors_info,
                    "\n{:<width$}{}",
                    "",
                    author,
                    width = pad + digit_difference(self.top_contribution(), author.contribution)
                )?;
            }
        }

        write!(f, "{authors_info}")
    }
}

#[typetag::serialize]
impl InfoField for AuthorsInfo {
    fn value(&self) -> String {
        self.to_string()
    }

    fn title(&self) -> String {
        let mut title: String = "Author".into();
        if self.authors.len() > 1 {
            title.push('s')
        }
        title
    }
}

#[cfg(test)]
mod test {
    use super::*;
    use crate::ui::text_colors::TextColors;
    use insta::assert_snapshot;
    use owo_colors::DynColors;

    #[test]
    fn test_display_author() {
        let author = Author::new(
            "John Doe".into(),
            Some("john.doe@email.com".into()),
            1500,
            2000,
            NumberSeparator::Plain,
        );

        assert_eq!(author.to_string(), "75% John Doe <john.doe@email.com> 1500");
    }

    #[test]
    fn test_display_author_with_no_email() {
        let author = Author::new("John Doe".into(), None, 1500, 2000, NumberSeparator::Plain);

        assert_eq!(author.to_string(), "75% John Doe 1500");
    }

    #[test]
    fn test_authors_info_title_with_one_author() {
        let author = Author::new(
            "John Doe".into(),
            Some("john.doe@email.com".into()),
            1500,
            2000,
            NumberSeparator::Plain,
        );

        let authors_info = AuthorsInfo {
            authors: vec![author],
        };

        assert_eq!(authors_info.title(), "Author");
    }

    #[test]
    fn test_authors_info_title_with_two_authors() {
        let author = Author::new(
            "John Doe".into(),
            Some("john.doe@email.com".into()),
            1500,
            2000,
            NumberSeparator::Plain,
        );

        let author_2 = Author::new(
            "Roberto Berto".into(),
            None,
            240,
            300,
            NumberSeparator::Plain,
        );

        let authors_info = AuthorsInfo {
            authors: vec![author, author_2],
        };

        assert_eq!(authors_info.title(), "Authors");
    }

    #[test]
    fn test_author_info_with_one_author() {
        let author = Author::new(
            "John Doe".into(),
            Some("john.doe@email.com".into()),
            1500,
            2000,
            NumberSeparator::Plain,
        );

        let authors_info = AuthorsInfo {
            authors: vec![author],
        };
        let colors = TextColors::new(&[], DynColors::Rgb(0xFF, 0xFF, 0xFF));
        let mut buffer = String::new();
        authors_info
            .write_styled(&mut buffer, false, &colors)
            .unwrap();

        assert_snapshot!(buffer);
    }

    #[test]
    fn test_author_info_with_two_authors() {
        let author = Author::new(
            "John Doe".into(),
            Some("john.doe@email.com".into()),
            1500,
            2000,
            NumberSeparator::Plain,
        );

        let author_2 = Author::new(
            "Roberto Berto".into(),
            None,
            240,
            300,
            NumberSeparator::Plain,
        );

        let authors_info = AuthorsInfo {
            authors: vec![author, author_2],
        };

        let colors = TextColors::new(&[], DynColors::Rgb(0xFF, 0xFF, 0xFF));
        let mut buffer = String::new();
        authors_info
            .write_styled(&mut buffer, false, &colors)
            .unwrap();

        assert_snapshot!(buffer);
    }
    #[test]
    fn test_author_info_alignment_with_three_authors() {
        let author = Author::new(
            "John Doe".into(),
            Some("john.doe@email.com".into()),
            1500,
            2000,
            NumberSeparator::Plain,
        );

        let author_2 = Author::new(
            "Roberto Berto".into(),
            None,
            240,
            300,
            NumberSeparator::Plain,
        );

        let author_3 = Author::new("Jane Doe".into(), None, 1, 100, NumberSeparator::Plain);

        let authors_info = AuthorsInfo {
            authors: vec![author, author_2, author_3],
        };

        let colors = TextColors::new(&[], DynColors::Rgb(0xFF, 0xFF, 0xFF));
        let mut buffer = String::new();
        authors_info
            .write_styled(&mut buffer, false, &colors)
            .unwrap();

        assert_snapshot!(buffer);
    }
}
