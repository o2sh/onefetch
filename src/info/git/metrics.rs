use super::sig::Sig;
use crate::cli::{CliOptions, NumberSeparator};
use crate::info::author::Author;
use crate::info::churn::FileChurn;
use anyhow::Result;
use gix::bstr::BString;
use gix::date::Time;
use std::collections::HashMap;

pub struct GitMetrics {
    pub authors_to_display: Vec<Author>,
    pub file_churns_to_display: Vec<FileChurn>,
    pub total_number_of_authors: usize,
    pub total_number_of_commits: usize,
    pub churn_pool_size: usize,
    pub time_of_most_recent_commit: gix::actor::Time,
    pub time_of_first_commit: gix::actor::Time,
}

impl GitMetrics {
    pub fn new(
        number_of_commits_by_signature: HashMap<Sig, usize>,
        number_of_commits_by_file_path: HashMap<BString, usize>,
        churn_pool_size: usize,
        time_of_first_commit: Option<Time>,
        time_of_most_recent_commit: Option<Time>,
        options: &CliOptions,
    ) -> Result<Self> {
        let total_number_of_commits = number_of_commits_by_signature.values().sum();
        let (authors_to_display, total_number_of_authors) = compute_authors(
            number_of_commits_by_signature,
            total_number_of_commits,
            options.info.number_of_authors,
            options.info.email,
            options.text_formatting.number_separator,
        );

        let file_churns_to_display = compute_file_churns(
            number_of_commits_by_file_path,
            options.info.number_of_file_churns,
            options.text_formatting.number_separator,
        );

        // This could happen if a branch pointed to non-commit object, so no traversal actually happens.
        let (time_of_first_commit, time_of_most_recent_commit) = time_of_first_commit
            .and_then(|a| time_of_most_recent_commit.map(|b| (a, b)))
            .unwrap_or_default();

        Ok(Self {
            authors_to_display,
            file_churns_to_display,
            total_number_of_authors,
            total_number_of_commits,
            churn_pool_size,
            time_of_first_commit,
            time_of_most_recent_commit,
        })
    }
}

fn compute_file_churns(
    number_of_commits_by_file_path: HashMap<BString, usize>,
    number_of_file_churns_to_display: usize,
    number_separator: NumberSeparator,
) -> Vec<FileChurn> {
    let mut number_of_commits_by_file_path_sorted = Vec::from_iter(number_of_commits_by_file_path);

    number_of_commits_by_file_path_sorted
        .sort_by(|(_, a_count), (_, b_count)| b_count.cmp(a_count));

    number_of_commits_by_file_path_sorted
        .into_iter()
        .map(|(file_path, nbr_of_commits)| {
            FileChurn::new(file_path.to_string(), nbr_of_commits, number_separator)
        })
        .take(number_of_file_churns_to_display)
        .collect()
}

fn compute_authors(
    number_of_commits_by_signature: HashMap<Sig, usize>,
    total_number_of_commits: usize,
    number_of_authors_to_display: usize,
    show_email: bool,
    number_separator: NumberSeparator,
) -> (Vec<Author>, usize) {
    let total_number_of_authors = number_of_commits_by_signature.len();
    let mut signature_with_number_of_commits_sorted: Vec<(Sig, usize)> =
        number_of_commits_by_signature.into_iter().collect();

    signature_with_number_of_commits_sorted.sort_by(|(sa, a_count), (sb, b_count)| {
        b_count.cmp(a_count).then_with(|| sa.name.cmp(&sb.name))
    });

    let authors: Vec<Author> = signature_with_number_of_commits_sorted
        .into_iter()
        .map(|(author, author_nbr_of_commits)| {
            Author::new(
                author.name.to_string(),
                if show_email {
                    Some(author.email.to_string())
                } else {
                    None
                },
                author_nbr_of_commits,
                total_number_of_commits,
                number_separator,
            )
        })
        .take(number_of_authors_to_display)
        .collect();
    (authors, total_number_of_authors)
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_compute_authors() {
        let mut number_of_commits_by_signature: HashMap<Sig, usize> = HashMap::new();
        number_of_commits_by_signature.insert(
            Sig {
                name: "John Doe".into(),
                email: "johndoe@example.com".into(),
            },
            10,
        );
        number_of_commits_by_signature.insert(
            Sig {
                name: "Jane Doe".into(),
                email: "janedoe@example.com".into(),
            },
            5,
        );
        number_of_commits_by_signature.insert(
            Sig {
                name: "Ellen Smith".into(),
                email: "ellensmith@example.com".into(),
            },
            50,
        );
        let total_number_of_commits = 15;
        let number_of_authors_to_display = 2;
        let show_email = false;
        let number_separator = NumberSeparator::Comma;

        let (authors, total_number_of_authors) = compute_authors(
            number_of_commits_by_signature,
            total_number_of_commits,
            number_of_authors_to_display,
            show_email,
            number_separator,
        );

        assert_eq!(total_number_of_authors, 3);
        assert_eq!(authors.len(), 2);
        assert_eq!(authors.get(0).unwrap().name, "Ellen Smith".to_string());
    }

    #[test]
    fn test_compute_file_churns() {
        let mut number_of_commits_by_file_path = HashMap::new();
        number_of_commits_by_file_path.insert("path/to/file1.txt".into(), 2);
        number_of_commits_by_file_path.insert("path/to/file2.txt".into(), 5);
        number_of_commits_by_file_path.insert("path/to/file3.txt".into(), 3);
        number_of_commits_by_file_path.insert("path/to/file4.txt".into(), 7);

        let number_of_file_churns_to_display = 3;
        let number_separator = NumberSeparator::Comma;
        let file_churns = compute_file_churns(
            number_of_commits_by_file_path,
            number_of_file_churns_to_display,
            number_separator,
        );

        assert_eq!(file_churns.len(), 3);
        assert_eq!(
            file_churns.get(0).unwrap().file_path,
            "path/to/file4.txt".to_string()
        );
        assert_eq!(file_churns.get(0).unwrap().nbr_of_commits, 7);
    }
}
