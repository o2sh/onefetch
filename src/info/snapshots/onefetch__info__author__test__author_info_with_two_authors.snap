---
source: src/info/author.rs
expression: buffer
---
[38;2;255;255;255;1mAuthors[0m[39;1m:[0m [39m75% John Doe <john.doe@email.com> 1500[0m
[39m         80% Roberto Berto 240[0m

